### DISCUSSION TITLE
adding call outs with a dedicated geom?

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#8] [Ideas]
Call outs, which allow you to tell more of a story about specific observations, can be tedious to add to plots.  

From what I've seen and experienced in doing this within ggplot2: the work flow may be something like... (annotate("label" , x  = ?, y = ?) + annotate("curve" or "segment"), and then many adjustments to label sit where you want it, and to make the label and link harmonious.  It's tempting to put the annotation at integers too, so it sits obviously in the grid, and it looks boxy when it should look higher-level observation comment that sits above it all.  An alternative is to export the plot to another tool ...  

So.... sketched-out geom_labellink [works in progress](https://evamaerey.github.io/mytidytuesday/2024-04-08-textcurve/textcurve.html).  Highlight below.  I think the logic of the callout placement matches better how we think about it (so less painful), and the label and link relative positions are handled *automatically* to be harmonious. 

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/fa03949a-aefd-42a7-949b-2ba1eebf54af)

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/1c56da83-661d-4da3-8217-cdf81e63eb3c)

If you don't provide a label, then the label that is printed is the value in the id column (which will be the same as the which_id character value).

You can leave off label direction too, so syntax can be super concise!   Default right now is that label lies SW of target, which should probably be replaced with something that looks at where target is relative to the rest of the data.

Overall placement of the label is based on the ranges of x and range of y, a direction and a proportion of the x y range.  There's also padding adjustment away from text and away from target - which also needs work.   Additionally, right now, a vector of inputs for which_id isn't working, just a single scalar.  

Arrowheads and curves development could make for a more sophisticated look and feel. 

This Stat extension is exciting because it's so audience- and storytelling- motivated!

Is there a ggrepel solution?  I'm more familiar with people using it to layer on a blanket of call outs, instead of a focused "this is one observation, and I'll write a little story about it and I will put that story right in this spot I selected in my plot panel space".

**Comment by teunbrand** [comment]
I think the {ggrepel} solution would look something like this:

``` r
library(tidyverse)

gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  ggrepel::geom_text_repel(
    aes(label = c("Chile" = "Chile", "United States" = "The US is here")[as.character(country)]),
    na.rm = TRUE, box.padding = unit(1, "cm"),
    point.padding = unit(5, "mm"),
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
```

![](https://i.imgur.com/IlRrsmC.png)

<sup>Created on 2024-04-12 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

I also think that [{ggpp}](https://github.com/aphalo/ggpp) has some labelling functions that are not entirely unlike the idea you presented, but I haven't tried any of them.

**Reply by EvaMaeRey** [reply]
Thank you for the example -  great auto-placement result!    The aes label definition syntax would be a challenge for me to produce.  I like the point.padding and box.padding argument names.  I thought of ggpp too.  A lot in that package and I do not have a good handle on what is there.   I guess with ggrepel, you wouldn't have the option to override and define your own label position?

**Reply by teunbrand** [reply]
I don't think hardcoded placements are possible in ggrepel, the repelling algorithm is recalculated every time the window is resized.

**Reply by aphalo** [reply]
Disabling repulsion and using nudging one can fake hardcoded placement. With 'ggpp' one can use `geom_text_s()` and position functions. Please, see my comment further down.

**Comment by EvaMaeRey** [comment]
another ggrepel ouput different fig height/width

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/2c7f5f68-9a0b-4666-ac88-77b0f358f59e)

**Comment by EvaMaeRey** [comment]
If we borrow thinking about facets, probably call outs for points are helpful for the same reasons.  From TLP's oneline workshop 'plot anything with ggplot2': 'Your mind will actually be able to reuse the knowledge it gained from reading the first panel.  Like it will understand how the plot should be read and it can apply that to all the other small multiples. So it's so cognitively it's much much easier for the head to wrap around small multiples.'  If we replace 'facet' with 'point', we'd make the argument that point callouts are helpful because they helps you interpret the representation of a single point (i.e. what is the observational unit?) and that understanding can be applied to all the other points.

**Comment by EvaMaeRey** [comment]
Advice from Matthew Blackwell that people love speaks directly to the value of explaining one point. 

'My best tip on how to give better quantitative presentations is to (a) use more plots and (b) build up your plots on multiple overlays, as in:

1. Just x-axis (explain it)
2. Add y-axis (explain it)
3. Add 1 data point (explain it)
4. Plot the rest of the data (explain it) '  https://twitter.com/matt_blackwell/status/991004129198854145

The approach here is to build up the plot - kind of an animation, because he's thinking about presentations.  But in a static plot a callout might stand in for the 'add one point, explain it step'.  Do 1. and 2 and 3 in title?

**Comment by EvaMaeRey** [comment]
And, callouts can be used like 'you-are-here' annotations are used on maps.  In other words, if the plot creator has a good understanding of their audience, then they can tell their audience how they fit into the picture.  i.e. the plot shows country characteristics and audience is Australian, doing a call out for Australia says 'you are here', look how we compare and this information is relevant to you.

**Reply by EvaMaeRey** [reply]
Malcolm Gladwell's (author of Outlier, and other non-fiction best sellers) endorsement of the 'you are here' communication: 'So a group of people if you ask them 'are in to data?' They'll say no, why would I be into data? In fact, are keenly interested in data when  it is about something that they're connected to'.... 'The notion that we're not interested in data is ... if you've had your eyes open ... from the earliest beginnings completely false.  the issue is do we have a connection to it.  And so all you need to do is create a connection.  People need a reason to look at a row of numbers.. ' 

He gives a couple of examples of data/chart that kids being extremely interest in - a seating chart based on grades from the last test  (historical practice), or running race results... 

'So you just need to take the time to set it up. So I will use charts and graphs in my writing all the time but only when I've told you a story as to why the chart or graph  is meaningful and how to read the chart or graph'

**Comment by EvaMaeRey** [comment]
I'm using a compute_panel so that ranges are computed across the entire range of the scatterplot; when you use the aes(id = my_unique_identifier) then you don't want the range to become defined by each point (0); When faceting, I'd like to have placement remain the same - defined by the entire constellation of data (all facets) not just by the panel.  So I thought I'd try to switch out to compute layer.  But this did not just work and I haven't used compute layer before so don't have a lot of context for doing things right. Just documenting thought process and naive expectations.   Back to compute_panel, but we see jump in positioning when moved to faceted plot (i.e. by continent) because range recomputes, which I don't think is a great behavior.

**Reply by aphalo** [reply]
'ggrepel' computes the repulsion in `draw_panel()`. Within `draw_panel()` it is possible to access the scale and get from it the range, which is set by 'ggplot2' based on all the panels, unless free scales are set in the facets. I do not remember out of the top of my head, if the scale can be accessed from within the `compute_panel()` function.

**Reply by EvaMaeRey** [reply]
Thank you for your insights and direction on this!

**Comment by aphalo** [comment]
Hi!

I fully agree with the usefulness of callouts, data labels, annotaions and insets. I think they were more common in the old times when draftment created plots using ink and pen. 

I am the author of 'ggpp'. Because I like to use these features, have emphasized them in package 'ggpp', which already a split off from my package 'ggpmisc'.

I give here a few hints of how to achieve good positioning with 'ggrepel'. 1) To ensure that the labels are repelled away from unlabelled points, keep all the rows in the data, but set the labels to be skipped to `""`.  2) What I think is the main problem with 'ggrepel' from a graphic design point of view is the randomness of the repulsion. One can limit it to x or y directions, but still this is not enough. What I discovered that works is to give the labels a small "push" in the desired direction at then let 'ggrepel' do its job. How I do this, is using nudging.  After using this approach manually, I tried to automate. In part I was experimenting, so there are quite a few `position` functions in 'ggpp', but I kept them because they all to be at least sometimes useful. By accident, or trial and error, I realized that position functions have access to `data`. This is crucial, because it makes possible to fit regression lines or 2D densities to the observations, and use the fitted line or local density of observations to set the nudging direction, and this small "push" in the "right" direction  "encourages" the repulsion to continue moving the labels in the same direction away from the cloud of observation, at least if there is space. Of course, one can also set an arbitrary line or point to guide the nudging direction. It was a surprise to myself how well this approach works in practice. 3) The repulsion force can be adjusted with arguments, and repuslon can be disabled completly is desired in the repulsive geoms from 'ggrepel'. 4) 'ggpp' provides some geoms with names ending in `_s` for segment, this segment joins the original position to the nudged or displaced position, like in the repulsive geoms, but with no repulsion ever applied. 5) The next problem I face was to add data labels or callouts to stack bar plots, and for this I added position functions such as `position_stack_nudge()` that can apply both displacements.

Although, off-topic, 'ggpp' also includes `geom_plot()`, `geom_table()` and `geom_grob()` that follow the same semantics as `geom_text()` and `geom_label()`, but accepting ggplots, data frames and grobs as labels. So, callouts can be small plots, small tables, or any grid graphical object. 

The repulsion code in 'ggrepel' is not something I properly undestand, but I think it should be possible to constrain the displacement to position the labels on a grid. I had not thought about this before reading this discussion.

I am happy to write some code examples. I will continue tomorrow as it is getting late over here in Finland.

**Reply by EvaMaeRey** [reply]
Very glad to have your contribution here! Code examples will be really valuable.  And these insights are really interesting -- I think I fully understand n. 1 (does it slow down rendering to have "" plotted would be one question I have).  I would like to more fully understand the other points.    Would you consider telling us more about your ggpp-ggrepel experiences in an informal talk at a ggplot2 extenders meetup?  That could basically be going through what you've outlined here, with a slide deck.    In general, I suspect that your insights and the utilities in ggpp could be useful to a wider range of users!  I'd certainly like to have a better working knowledge of them, and glad to be hearing more about it thus far.

**Reply by EvaMaeRey** [reply]
I asked some folks who I regularly see using annotation on social media about their go-to methods:  https://twitter.com/EvaMaeRey/status/1781422041771462830

**Comment by aphalo** [comment]
@EvaMaeRey I am glad to be able to contribute! Thanks for the invitation to talk!
I give here some examples using only 'ggrepel', starting from Teun's code. As you already noticed, repulsion in 'ggrepel' uses a random displacement, so running the same code can result in a different position of labels in a plot. In addition, as the repulsion is applied during rendering, the position of repulsed labels depends on the size used to render the plot. Repulsion can disabled or restricted to one direction. Force can be adjusted as well and maximum number of overlaps at which the repulsion algorithm gives up.

By setting labels to `""`, 'ggrepel' does not attempt to change their positions, however, it repulses other labels away from the unlabelled points. Is this slower? yes to some extent, but not much, I think. It is at least much faster that repulsing actuals labels for every point. (If the labels are set to `NA` the whole row is removed from `data` before the plot is rendered. This is why `""` was chosen as the marker for no label.)

While position functions from 'ggplot2' only modify `x` and `y`, the `position_nudge_repel()` retains in data a copy of the original values of `x` and `y`.  (I copied this approach in 'ggpp', but using a slightly different naming scheme, so that the position functions in 'ggpp' could remain compatible with normal geoms from 'ggplot2' and extensions).

``` r
library(tidyverse)
library(ggrepel)

# sometimes labels overlap the points, depending on the random numbers seed
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_text_repel(
    aes(label = c("Chile" = "Chile", "United States" = "The US is here")[as.character(country)]),
    na.rm = TRUE, box.padding = unit(1, "cm"),
    point.padding = unit(5, "mm"),
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
```

![](https://i.imgur.com/pWukYSI.png)

``` r

# labels never overlap the points as long as there is space
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_text_repel(
    aes(label = ifelse(country %in% c("Chile", "United States"),
                       c("Chile" = "Chile", "United States" = "The US is here")[as.character(country)],
                       "")),
    na.rm = TRUE, box.padding = unit(1, "cm"),
    point.padding = unit(5, "mm"),
    arrow = arrow(length = unit(2, "mm"), type = "closed"),
    max.overlaps = 30
  )
```

![](https://i.imgur.com/8gxXPx4.png)

``` r

# In this case we can restrict repulsion to the vertical axis
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_text_repel(
    aes(label = ifelse(country %in% c("Chile", "United States"),
                       c("Chile" = "Chile", "United States" = "The US is here")[as.character(country)],
                       "")),
    direction = "y",
    na.rm = TRUE, box.padding = unit(1, "cm"),
    point.padding = unit(5, "mm"),
    arrow = arrow(length = unit(2, "mm"), type = "closed"),
    max.overlaps = 30
  )
```

![](https://i.imgur.com/VUXdVwq.png)

``` r

# Using nudging to start movement upwards
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_text_repel(
    aes(label = ifelse(country %in% c("Chile", "United States"),
                       c("Chile" = "Chile", "United States" = "The US is here")[as.character(country)],
                       "")),
    nudge_y = 3,
    direction = "y",
    na.rm = TRUE, box.padding = unit(1, "cm"),
    point.padding = unit(5, "mm"),
    arrow = arrow(length = unit(2, "mm"), type = "closed"),
    max.overlaps = 30
  )
```

![](https://i.imgur.com/5haifCA.png)

``` r

# Disabling repulsion
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  ggrepel::geom_text_repel(
    aes(label = ifelse(country %in% c("Chile", "United States"),
                       c("Chile" = "Chile", "United States" = "The US is here")[as.character(country)],
                       "")),
    nudge_y = 4,
    direction = "y",
    force = 0,
    force_pull = 0,
    na.rm = TRUE, box.padding = unit(1, "cm"),
    point.padding = unit(5, "mm"),
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
```

![](https://i.imgur.com/S1A0wIp.png)

<sup>Created on 2024-05-01 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

In this case, the best approach seems to be the 4th: nudging + repulsion along _y_.

Tomorrow I should have a bit more time, and I will post here some variations using 'ggpp'.

**Comment by aphalo** [comment]
Hi. 

Here are some examples with 'ggpp' and with 'ggrepel'. I do not think any automatic approach can provide plots as nice as those than be obtained by manual placement. One important difference between 'ggrepel' and 'ggpp' is that `geom_text_repel()` avoids clipping of labels at the edge of the plotting area, while `geom_text_s()` behaves like `geom_text()` and may clip part  of labels. Of course, the solution is to manually expand the axis limits.

Padding 'ggpp' does not support 'grid' units, differemtly to 'ggrepel'.

``` r
library(tidyverse)
library(ggpp)
#> Registered S3 methods overwritten by 'ggpp':
#>   method                  from   
#>   heightDetails.titleGrob ggplot2
#>   widthDetails.titleGrob  ggplot2
#> 
#> Attaching package: 'ggpp'
#> The following object is masked from 'package:ggplot2':
#> 
#>     annotate
library(ggrepel)

# Using 'ggpp' ------------------------------------------------------------

# I label the same two countries as in the original example

# Using non-repulsive geom
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_text_s(
    aes(label = c("Chile" = "Chile", "United States" = "The US is here")[as.character(country)]),
    nudge_y = 4,
    na.rm = TRUE, 
    box.padding = 0.3,
    point.padding = 0.3,
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
```

![](https://i.imgur.com/Oy2bAJa.png)

``` r

# Justification changes the attachment of the arrow
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_text_s(
    aes(label = c("Chile" = "Chile", "United States" = "The US is here")[as.character(country)]),
    nudge_y = 4, nudge_x = 1000,
    hjust = "left",
    na.rm = TRUE, 
    box.padding = 0.3,
    point.padding = 0.3,
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
```

![](https://i.imgur.com/fieziT9.png)

``` r

# Nudge can be a vector (it will be recycled if necessary)
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_text_s(
    aes(label = c("Chile" = "Chile", "United States" = "The US is here")[as.character(country)]),
    nudge_y = c(-4, 4), nudge_x = 1000,
    hjust = "left",
    na.rm = TRUE, 
    box.padding = 0.3,
    point.padding = 0.3,
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
```

![](https://i.imgur.com/7IZnOik.png)

``` r

# Nudging to a target position
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
   geom_text_s(
    aes(label = c("Chile" = "Chile", "United States" = "The US is here")[as.character(country)]),
    position = position_nudge_to(y = 82),
    hjust = "left",
    na.rm = TRUE, 
    box.padding = 0.3,
    point.padding = 0.3,
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
```

![](https://i.imgur.com/Qf2uwpu.png)

``` r

# Here I label more countries than in the original example

some.countries <- c("Chile" = "Chile",
                    "Argentina" = "Argentina",
                    "Paraguay" = "Paraguay",
                    "Bolivia" = "Bolivia",
                    "Peru" = "Perú",
                    "Somalia" = "Somalia",
                    "United States" = "The US is here")

# Nudging away from the cloud of points, which does not work well without repulsion
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_point(aes(alpha = country %in% names(some.countries)), show.legend = FALSE) +
  geom_text_s(
    aes(label = some.countries[as.character(country)]),
    position = position_nudge_line(y = 4, x =1000, method = "spline"),
    hjust = "outward",
    na.rm = TRUE, 
    box.padding = 0.3,
    point.padding = 0.3,
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
#> Warning: Using alpha for a discrete variable is not advised.
```

![](https://i.imgur.com/tkXI4Oj.png)

``` r


# Nudging away from the cloud of points, increasing nudging helps
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_point(aes(alpha = country %in% names(some.countries)), show.legend = FALSE) +
  geom_text_s(
    aes(label = some.countries[as.character(country)]),
    position = position_nudge_line(y = 6, x =2500, method = "spline"),
    na.rm = TRUE, 
    box.padding = 0.3,
    point.padding = 0.3,
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
#> Warning: Using alpha for a discrete variable is not advised.
```

![](https://i.imgur.com/uEpD7FT.png)

``` r

# Nudging away from the cloud of points, increasing nudging and setting hjust
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_point(aes(alpha = country %in% names(some.countries)), show.legend = FALSE) +
  geom_text_s(
    aes(label = some.countries[as.character(country)]),
    position = position_nudge_line(y = 6, x =2500, method = "spline"),
    hjust = "outward",
    na.rm = TRUE, 
    box.padding = 0.3,
    point.padding = 0.3,
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
#> Warning: Using alpha for a discrete variable is not advised.
```

![](https://i.imgur.com/O40U0oK.png)

``` r

# Nudging away from the cloud of points, setting nudging direction to "split"
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_point(aes(alpha = country %in% names(some.countries)), show.legend = FALSE) +
  geom_text_s(
    aes(label = some.countries[as.character(country)]),
    position = position_nudge_line(y = 4, x = 1000, method = "spline",
                                   direction = "split"),
    na.rm = TRUE, 
    box.padding = 0.3,
    point.padding = 0.3,
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
#> Warning: Using alpha for a discrete variable is not advised.
```

![](https://i.imgur.com/tcxapGR.png)

``` r

# Using 'ggpp' + 'ggrepel' ------------------------------------------------

# The examples above tend to work, but there is no guarantee of no overlaps,
# so combining them with repulsion is safer. The more dispersion there is
# in the data points the more necessary repulsion becomes.

# I set max.overlaps = 30 so that 'ggrepel' does not give up too early

# Nudging away based on a spline fitted to the cloud of points
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_point(aes(alpha = country %in% names(some.countries)), show.legend = FALSE) +
  geom_text_repel(
    aes(label = ifelse(country %in% names(some.countries),
                       some.countries[as.character(country)],
                       "")),
    position = position_nudge_line(y = 5, x = 3000, method = "spline"),
    na.rm = TRUE, 
    arrow = arrow(length = unit(2, "mm"), type = "closed"),
    max.overlaps = 30
  )
#> Warning: Using alpha for a discrete variable is not advised.
```

![](https://i.imgur.com/Be4D7LL.png)

``` r

# Nudging away based on a spline fitted to the cloud of points with repulsion and hjust
# The effect of justification is different in the geoms from 'ggrepel' and 'ggpp'!
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_point(aes(alpha = country %in% names(some.countries)), show.legend = FALSE) +
  geom_text_repel(
    aes(label = ifelse(country %in% names(some.countries),
                       some.countries[as.character(country)],
                       "")),
    position = position_nudge_line(y = 5, x = 3000, method = "spline"),
    hjust = "outward", # <- added
    na.rm = TRUE, 
    arrow = arrow(length = unit(2, "mm"), type = "closed"),
    max.overlaps = 30
  )
#> Warning: Using alpha for a discrete variable is not advised.
```

![](https://i.imgur.com/kq6Q3nV.png)

``` r

# Nudging away from the cloud of points with nudging with direction away from
# fitted spline.
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_point(aes(alpha = country %in% names(some.countries)), show.legend = FALSE) +
  geom_text_repel(
    aes(label = ifelse(country %in% names(some.countries),
                       some.countries[as.character(country)],
                       "")),
    position = position_nudge_line(y = 5, x = 3000, method = "spline",
                                  direction = "split"), # <-- added
    hjust = "outward",
    na.rm = TRUE, 
    max.overlaps = 50,
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
#> Warning: Using alpha for a discrete variable is not advised.
```

![](https://i.imgur.com/v5uhdOn.png)

``` r

# Nudging away from the cloud of points with repulsion, a different option for
# direction.
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_point(aes(alpha = country %in% names(some.countries)), show.legend = FALSE) +
  geom_text_repel(
    aes(label = ifelse(country %in% names(some.countries),
                       some.countries[as.character(country)],
                       "")),
    position = position_nudge_line(y = 5, x = 3000, method = "spline",
                                   direction = "automatic"), # <-- added
    na.rm = TRUE, 
    max.overlaps = 50,
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
#> Warning: Using alpha for a discrete variable is not advised.
```

![](https://i.imgur.com/AdOeeSQ.png)

``` r

# Nudging away from the cloud of points with repulsion, adding a minimum
# distance from the fitted line for the nudging that is 1.5 times the
# nudging of observations.
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_point(aes(alpha = country %in% names(some.countries)), show.legend = FALSE) +
  geom_text_repel(
    aes(label = ifelse(country %in% names(some.countries),
                       some.countries[as.character(country)],
                       "")),
    position = position_nudge_line(y = 5, x = 3000, method = "spline",
                                   direction = "automatic",
                                   line_nudge = 1.5),
    na.rm = TRUE, 
    max.overlaps = 50,
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
#> Warning: Using alpha for a discrete variable is not advised.
```

![](https://i.imgur.com/JrW8Hta.png)

``` r

# Nudging away from the cloud of points with repulsion, adding a minimum
# distance from the fitted line that is 2 times the nudging for individual
# observations, while reducing the amount of nudging for observations.
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_point(aes(alpha = country %in% names(some.countries)), show.legend = FALSE) +
  geom_text_repel(
    aes(label = ifelse(country %in% names(some.countries),
                       some.countries[as.character(country)],
                       "")),
    position = position_nudge_line(y = 3, x = 2000, method = "spline",
                                   direction = "split",
                                   line_nudge = 2),
    na.rm = TRUE, 
    max.overlaps = 50,
    min.segment.length = 0,
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
#> Warning: Using alpha for a discrete variable is not advised.
```

![](https://i.imgur.com/yqr3Fpj.png)

<sup>Created on 2024-05-01 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
@aphalo  This is amazing!  Thank you for so much detail and the examples.   I'm engaging w/ code, going through plot by plot.  I think I'll do a little bit of a rewrite to try to isolate the new thing/strategy that you are communicating with each figure - which I enjoy... :-)  The position functions seem really interesting!

**Comment by aphalo** [comment]
@EvaMaeRey Feel free to ask if you have any questions. This thread is forcing me to rethink things... which is very useful for improving 'ggpp'.

I haven't looked at your code, so I may be repeating what is there. None of `geom_text()`, `geom_text_repel()` or `geom_text_s()` support automatic wrapping of the text, but they all respect newline characters, so wrapping is possible with all of them. In the examples below I used base R functions for kludging a function `wrap_label()` able to do the wrapping of the individual strings in a character vector. (Of course, the tidyverse could be used instead of base R.)

I also adjusted (played with) nudging.

``` r
library(tidyverse)
library(ggrepel)
library(ggpp)
#> Registered S3 methods overwritten by 'ggpp':
#>   method                  from   
#>   heightDetails.titleGrob ggplot2
#>   widthDetails.titleGrob  ggplot2
#> 
#> Attaching package: 'ggpp'
#> The following object is masked from 'package:ggplot2':
#> 
#>     annotate

wrap_label <- function(x, width) {
    sapply(X = x,
           FUN = function(x) {
             paste(strwrap(x, width = width, prefix = "\n", initial = ""), collapse = "")
           })
}

country.labels <- c("Chile" = "República de Chile", "United States" = "United States of America")

# Using a repulsive geom from 'ggrepel'
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_text_repel(
    aes(label = ifelse(country %in% names(country.labels),
                       wrap_label(country.labels, width = 15)[as.character(country)],
                       "")),
    hjust = 0,
    nudge_y = c(-6, 4),
    nudge_x = 1000,
    na.rm = TRUE, 
    box.padding = 0.3,
    point.padding = 0.3,
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
```

![](https://i.imgur.com/DTnSlKq.png)

``` r


# Using a non-repulsive geom 'ggpp'
gapminder::gapminder |>
  filter(year == 2002) |>
  ggplot() +
  aes(gdpPercap, lifeExp) +
  geom_point(colour = "darkgrey") +
  geom_text_s(
    aes(label = wrap_label(country.labels, width = 15)[as.character(country)]),
    hjust = 0,
    nudge_y = c(-3, 3),
    nudge_x = 1000,
    na.rm = TRUE, 
    box.padding = 0.3,
    point.padding = 0.3,
    arrow = arrow(length = unit(2, "mm"), type = "closed")
  )
```

![](https://i.imgur.com/diMxyOQ.png)

<sup>Created on 2024-05-02 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by teunbrand** [comment]
Perhaps also worth pointing out here is that Thomas Lin Pedersen is working on a new markdown rendering guts if you want text wrapping.
The package is over at https://github.com/thomasp85/marquee and in particular `geom_marquee()` might be worth giving a spin.

**Reply by EvaMaeRey** [reply]
Looks exciting! I've noticed some announcements.  I think too, that a filter to one case, then nudge approach w/ ggrepel is interesting for the journalist - expository type label that is my motivation.  So there's basically no repulsion (I think), so the label sits on top of the point, then nudge to whereever you want?  But desirable would be npc nudging... Wrap some functions around that to specify angle direction and npc length rather than npc x and npc y nudge.   Then I think we'd really be working in a similar way to how I think most people intuitively would describe adding a call out.   
![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/2aae682a-f2ae-46c1-b496-b67510450c83)
